### PR TITLE
Add Notation data support in subpackets

### PR DESCRIPTION
**Pull request by wussler**
see https://github.com/ProtonMail/crypto/issues/27

### CODE REVIEW

**Comment by YarmoM** [comment]
Dear maintainers, will this get reviewed?

**Comment by wiktor-k** [comment]
Hi folks! :wave: 

Some people brought it to my attention that notation support is critical to their use case, e.g. writing [Keyoxide](https://keyoxide.org/)-compatible tools (Keybase-like decentralized bidirectional proof systems) in Go.

If conflicts are the only thing preventing the merge I know some people even have fixed that in their forks (@omz13) and could raise a fix-up PR if that's the only concern.

Would that be an acceptable resolution? Thanks for your time! CC: @twiss, @wussler.

(I'll use this occasion to plug an ad that [Sequoia PGP](https://sequoia-pgp.org/) has a nice and full support for notations! :sweat_smile: haha)

**Comment by wussler** [comment]
Hi @wiktor-k, thank you for your interest. 

We've seen there is quite some push for this feature, and will take care of merging this PR!

**Comment by wiktor-k** [comment]
Thank you, very, *very* much! :bow: Have a nice day!

**Comment by wiktor-k** [comment]
Good point @twiss, I [should have remembered that](https://github.com/openpgpjs/openpgpjs/commit/82799390ded7bd023555f481fc6992a560e7924b)! :sweat_smile:

**Comment by twiss** [comment]
Right :D Thanks for pushing this forward in all our implementations ^.^

**Comment by omz13** [comment]
The `Notations` structure does not contain any exportable fields... should they be exported? Just thinking about this because I'm dumping some entities as JSON data and the Notations entry is there but an empty object because nothing to export (even though there's some notation data present).

**Comment by twiss** [comment]
> IsHumanReadable?
> 
> > Critical      bool
> 
> IsCritical?

Yeah, maybe that's better :+1:

**Comment by wiktor-k** [comment]
I just noticed (for the record) that my old commit contains quite a few examples that could be used in unit tests for critical notations here:

https://github.com/openpgpjs/openpgpjs/commit/82799390ded7bd023555f481fc6992a560e7924b#diff-c789f741e120e8bd74404f74450907c81bcc9918eb3de78037b04b15552290e7R470-R637

**Comment by twiss** [comment]
I've added `config.KnownNotations`, a `map[string]bool` indicating which notation names are known (i.e. allowed for critical notations). This is then checked when reading a message / verifying a signature, both for the signature itself and all relevant key binding signatures. Let me know if anyone sees an issue with this API!

@wiktor-k thanks for the link to the test case, I've added it here as well :)

File: openpgp/packet/notation.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// Notation type represents a Notation Data subpacket
+// see https://tools.ietf.org/html/rfc4880#section-5.2.3.16
+type Notation struct {
+	Name string
+	Value []byte
+	IsCritical bool
+	IsHumanReadable bool
+}
+
+func (not *Notation) getData() []byte {
```
Comment (marinthiercelin) [inline openpgp/packet/notation.go:12]: I'd find it easier to read if the variable was not called 'not'

File: openpgp/packet/notation.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	if not.IsHumanReadable {
+		data[0] = 0x80
+	}
+
+	data[4] = byte(nameLen >> 8)
+	data[5] = byte(nameLen)
+	data[6] = byte(valueLen >> 8)
+	data[7] = byte(valueLen)
+
+	data = append(data, nameData...)
```
Comment (marinthiercelin) [inline openpgp/packet/notation.go:27]: I think append would not work as intended here.
Data is already set to have length = 8 + nameLen + valueLen, and append will write after those bytes, not overwrite them.
I think what we want is to use `copy` here.
See https://github.com/ProtonMail/go-crypto/pull/144

File: openpgp/packet/notation.go (hunk 1 above)
Comment (twiss) [reply openpgp/packet/notation.go]: :yum: Hmhm :+1: Renamed it to `notation` .

File: openpgp/packet/signature.go [hunk 3]
```diff
@@ -367,6 +369,24 @@ func parseSignatureSubpacket(sig *Signature, subpacket []byte, isHashed bool) (r
 		}
 		sig.IssuerKeyId = new(uint64)
 		*sig.IssuerKeyId = binary.BigEndian.Uint64(subpacket)
+	case notationDataSubpacket:
+		// Notation data, section 5.2.3.16
+		nameLength := uint32(subpacket[4])<<8 | uint32(subpacket[5])
```
Comment (marinthiercelin) [inline openpgp/packet/signature.go:374]: Shouldn't we check that subpacket is larger than 8 bytes before accessing ?

**Review by marinthiercelin — APPROVED** [review]

**Comment by wiktor-k** [comment]
Thanks everyone! 👏 Great to see this finally being in! 👍

Have a great evening! 👋

**Comment by twiss** [comment]
Thanks, you too :)

**Comment by omz13** [comment]
Many many many thanks!

**Pull request merged**